### PR TITLE
add security tests for cardano dapp support

### PR DESCRIPTION
**Pull request by kdenhartog**
These were mostly tests generated during static code analaysis of the cardano dapp functionality. I wanted to make sure that I understood what the code was doing by having an agent add these tests and then I reviewed them for functional cover of edge cases that I felt could impact the security of the wallet.

Related to https://github.com/brave/reviews/issues/2000

### CODE REVIEW

File: components/brave_wallet/browser/cardano/cardano_wallet_service_unittest.cc [hunk 1]
```diff
@@ -175,6 +175,130 @@ TEST_F(CardanoWalletServiceUnitTest, GetChangeAddress) {
...
+TEST_F(CardanoWalletServiceUnitTest, NoAddressMethod_ReturnsMultipleAddresses) {
+  SetupCardanoAccount();
+
+  // GetUsedAddresses should NOT return multiple addresses
+  auto used_addresses = cardano_wallet_service_->GetUsedAddresses(account_id());
+  EXPECT_EQ(used_addresses.size(), 1u)
+      << "GetUsedAddresses returned " << used_addresses.size()
+      << " addresses. If this is intentional (Issue #45278), update this test "
+         "and review privacy implications in "
+         "task-1.8-address-correlation-risk.md";
```
Comment (kdenhartog) [inline components/brave_wallet/browser/cardano/cardano_wallet_service_unittest.cc]: todo remove these

File: components/brave_wallet/browser/cardano/cardano_wallet_service_unittest.cc [hunk 2]
```diff
@@ -228,4 +352,180 @@ TEST_F(CardanoWalletServiceUnitTest, CreateAndSignCardanoTransaction) {
       cardano_test_rpc_server_->captured_raw_tx());
 }
 
+TEST_F(CardanoWalletServiceUnitTest,
+       SignAndPostTransaction_AccountSwitch_TOCTOU) {
```
Comment (kdenhartog) [inline components/brave_wallet/browser/cardano/cardano_wallet_service_unittest.cc]: this test kind of feels like junk - I ended up manually verifying this which is where I discovered that the sync calls hold the transactions if multiple are awaiting user interaction. I'll probably drop this unless you see a good reason to keep @supermassive

File: components/brave_wallet/browser/cardano/cardano_provider_impl_unittest.cc [hunk 3]
```diff
@@ -447,4 +447,314 @@ TEST_F(CardanoProviderImplUnitTest, EnableFails_OnWalletUnlock_TabNotActive) {
...
+
+  TestFuture<mojom::CardanoProviderErrorBundlePtr> future;
+  mojo::Remote<mojom::CardanoApi> api;
+  provider()->Enable(api.BindNewPipeAndPassReceiver(), future.GetCallback());
+
+  auto error = future.Take();
+  EXPECT_FALSE(error);
+}
+
+TEST_F(CardanoProviderImplUnitTest, Enable_UnlockRetainsCapturedOrigin) {
```
Comment (kdenhartog) [inline components/brave_wallet/browser/cardano/cardano_provider_impl_unittest.cc:514]: this isn't really needed since the response calls back to the origin via the promise, but is kind of useful for the defense in depth check

Comment (kdenhartog) [reply components/brave_wallet/browser/cardano/cardano_provider_impl_unittest.cc:514]: decided to keep this since it's useful for defense in depth

File: components/brave_wallet/browser/cardano/cardano_wallet_service_unittest.cc (hunk 2 above)
Comment (kdenhartog) [reply components/brave_wallet/browser/cardano/cardano_wallet_service_unittest.cc]: deleted

File: components/brave_wallet/browser/cardano/cardano_wallet_service_unittest.cc (hunk 1 above)
Comment (kdenhartog) [reply components/brave_wallet/browser/cardano/cardano_wallet_service_unittest.cc]: removed

**Comment by kdenhartog** [comment]
Converting this back to draft until I've had a chance to clean this up better and make sure the code is maintainable by the crypto-wallet-core team. The primary purpose of generating these tests were for me to gain confidence in edge cases of the ADA implementation, but I didn't focus on making sure they were runnable in CI or production ready.

File: components/brave_wallet/browser/cardano/cardano_cip30_serializer_unittest.cc [hunk 4]
```diff
@@ -124,4 +124,112 @@ TEST(CardanoCip30SerializerTest, SerializeUtxos) {
...
+      "d9ef8dcd983c6fe996d5029e010e224bec191d0f63ff695cdab046abfd79dfbd",
+      &tx_bytes);
+  base::SpanWriter(base::span(output.tx_hash)).Write(tx_bytes);
+  output.address_to = *CardanoAddress::FromString(
+      "addr1qyx2zscdearcexdktcgq6g27jkyff65dw82w6catczfwxz2qjy"
+      "nwf42y3c7ejrrekj5r2fh6kx5m9gcrmywpqxw3np5qjeh38p");
+  utxos.push_back(std::move(output));
+
+  auto result = CardanoCip30Serializer::SerializeUtxos(utxos);
+  EXPECT_TRUE(result.has_value());
```
Comment (cypt4) [inline components/brave_wallet/browser/cardano/cardano_cip30_serializer_unittest.cc:143]: We should use valid result test vectors, not just testing result has value
